From d5e113cb4a20e43482f89222863e6deaa65da192 Mon Sep 17 00:00:00 2001
From: James Bottomley <jejb@linux.ibm.com>
Date: Sat, 28 Nov 2020 15:27:21 -0800
Subject: [PATCH 775/921] [Backport] scsi: ses: Don't attach if enclosure has
 no components

stable inclusion
from stable-v5.10.173
commit d68937dfc73ee7f61cf3424fa3225be93cacaa00
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6PMQI

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=d68937dfc73ee7f61cf3424fa3225be93cacaa00

--------------------------------

commit 3fe97ff3d94934649abb0652028dd7296170c8d0 upstream.

An enclosure with no components can't usefully be operated by the driver
(since effectively it has nothing to manage), so report the problem and
don't attach. Not attaching also fixes an oops which could occur if the
driver tries to manage a zero component enclosure.

[mkp: Switched to KERN_WARNING since this scenario is common]

Link: https://lore.kernel.org/r/c5deac044ac409e32d9ad9968ce0dcbc996bfc7a.camel@linux.ibm.com
Cc: stable@vger.kernel.org
Reported-by: Ding Hui <dinghui@sangfor.com.cn>
Signed-off-by: James Bottomley <James.Bottomley@HansenPartnership.com>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 drivers/scsi/ses.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/scsi/ses.c b/drivers/scsi/ses.c
index cdf126bd0ffb..8a048988cff7 100644
--- a/drivers/scsi/ses.c
+++ b/drivers/scsi/ses.c
@@ -703,6 +703,12 @@ static int ses_intf_add(struct device *cdev,
 		    type_ptr[0] == ENCLOSURE_COMPONENT_ARRAY_DEVICE)
 			components += type_ptr[1];
 	}
+
+	if (components == 0) {
+		sdev_printk(KERN_WARNING, sdev, "enclosure has no enumerated components\n");
+		goto err_free;
+	}
+
 	ses_dev->page1 = buf;
 	ses_dev->page1_len = len;
 	buf = NULL;
-- 
2.17.1

